Add tutorial on how to estimate a Cox Proportional Hazards Model in glum

MatthiasSchmidtblaicherQC opened the pull request:
And: Add link to Matt Mills' tutorial on fitting penalized splines in glum.

Checklist
* [ ] Added a `CHANGELOG.rst` entry

stanmart commented on docs/tutorials/cox_model/cox_model.ipynb:
```diff
@@ -0,0 +1,893 @@
+    "\n",
+    "## 1. Equivalence Between the Cox Likelihood and a Profile Poisson Likelihood<a class=\"anchor\"></a>\n",
+    "\n",
+    "In the Cox model, the rate of event occurrence, $\\lambda(t,x_i)$, factorizes nicely into a linear predictor $\\eta_i=\\sum_k \\beta_k x_{ik}$ that depends on individual $i$'s characteristics but not on time $t$ and a baseline hazard $\\lambda_0$ that depends only on time, $\\lambda(t,x_i)=\\lambda_0(t)\\exp(\\eta_i)$ (the proportional hazards assumption). The partial log-likelihood of $\\et ...
```
```suggestion
    "In the Cox model, the rate of event occurrence, $\\lambda(t,x_i)$, factorizes nicely into a linear predictor $\\eta_i=\\sum_k \\beta_k x_{ik}$ that depends on individual $i$'s characteristics but not on time $t$, and a baseline hazard $\\lambda_0$ that depends only on time: $\\lambda(t,x_i)=\\lambda_0(t)\\exp(\\eta_i)$. This is known as the proportional hazards assumption). The partial log-likelihood of $\\eta_i$ is\n",
```

stanmart commented on docs/tutorials/cox_model/cox_model.ipynb:
```diff
@@ -0,0 +1,893 @@
+    "\n",
+    "## 2. Estimating a Cox Model in Glum<a class=\"anchor\"></a>\n",
+    "\n",
+    "We now show that a Poisson approach in `glum` yields the same parameter estimates as a Cox model. For the latter, we use the [lifelines](https://github.com/CamDavidsonPilon/lifelines) library. We also take the dataset from lifelines, which is from an RCT on recidivism for 432 convicts released from Maryland state prisons with first arrest after release as event. We first load imports and the ...
```
```suggestion
    "We now show that a Poisson approach in `glum` yields the same parameter estimates as a Cox model. For the latter, we use the [lifelines](https://github.com/CamDavidsonPilon/lifelines) library. We also take the dataset from lifelines, which is from an RCT on recidivism for 432 convicts released from Maryland state prisons with first arrest after release as event. We first load imports and the dataset. The dataset has one row per convict, with two outcome columns: the `week` until which the observation lasts and `arrest`, which indicates whether an arrest event happened or not (censoring)."
```

stanmart commented on docs/tutorials/cox_model/cox_model.ipynb:
```diff
@@ -0,0 +1,893 @@
+   "source": [
+    "## Footnotes\n",
+    "<span id=\"fn1\"> 1</span>:\n",
+    "The Cox model assumes that at most one individual has an event at any time (\"no ties\").\n",
```
Maybe mention here that there are various ways to deal with this? (e.g., exact - and expensive - calculations, efron approximation (reasonably accurate), breslow approximation (fastest). Also, in the case of many ties, an inherently discrete survival model might be the best option.

stanmart commented on docs/tutorials/cox_model/cox_model.ipynb:
```diff
@@ -0,0 +1,893 @@
+    "\n",
+    "[2] Whitehead, J., 1980. Fitting Cox’s regression model to survival data using GLIM. _Journal of the Royal Statistical Society Series C: Applied Statistics_, 29(3), pp.268-275.\n",
+    "\n",
+    "[3] Zhong, C. and Tibshirani, R., 2019. Survival analysis as a classification problem. arXiv preprint arXiv:1909.11171."
```
They do use the same dataset-transformation (which they call stacking) idea, but they rely on a binomial model instead of Poisson and say that it is a good enough approximation. So relevant, but I would not say it's the same thing.

MatthiasSchmidtblaicherQC replied:
Right. I removed that reference.

stanmart reviewed (approved):
Thank you Matthias, I like this tutorial a lot!

Maybe we could also mention that the model we estimate is conceptually a piecewise constant hazard model, but because we have a different constant for each event time, it becomes essentially non-parametric. As a corollary, if one wants to reduce the number of parameters, they can opt for merging some categories and estimating fewer "constants".

MatthiasSchmidtblaicherQC commented:
> Thank you Matthias, I like this tutorial a lot!
> 
> Maybe we could also mention that the model we estimate is conceptually a piecewise constant hazard model, but because we have a different constant for each event time, it becomes essentially non-parametric. As a corollary, if one wants to reduce the number of parameters, they can opt for merging some categories and estimating fewer "constants".

Thanks! I incorporated your suggestion and added a small conclusion section describing how one might want to depart from the Cox model. Feel free to have another look :).
